### PR TITLE
Adding inline transactions in payload to reduce latency

### PR DESCRIPTION
**Pull request by vusirikala**
### Description
With quorum store, batch proposer broadcasts a batch, other validators send back a vote on the batch, and the batch proposer than sends out the proof of store on the batch. Block proposer only includes batches with proof of store inside its proposal. Quroum store leads to a 3 hop latency between batch proposal and block proposal. To reduce this latency, we are allowing the proposal to contain a limited number of inline transactions with a proof of store. This means, the block proposer doesn't have to wait 3 hops to use the batch inside the proposal. 

### Test Plan
100 node sweep test on forge with this PR - https://github.com/aptos-labs/aptos-core/actions/runs/8166381194
100 node sweep test on forge before this PR - https://github.com/aptos-labs/aptos-core/actions/runs/8026314881
The results show that this PR gives a 2 hop (200 ms) latency reduction in case of low loads.

### CODE REVIEW

**Comment by bchocho** [comment]
Summarizing an offline conversation with @vusirikala:
* We'll need to have the inline batches also added to PayloadFilter.
* We need to be careful about a malicious validator lying about batches here (as there is no quorum signature yet). It should be ok that we verify each inline batch digest matches with the txn hashes. And then on filtering we'll have to make sure the digest again matches before filitering the actual proof.

In addition, I'm not sure we want to have inline batches potentially with every block. It's extra work. Probably just want inline batches when the block is smaller than a threshold.

File: consensus/consensus-types/src/common.rs [hunk 1]
```diff
@@ -298,6 +348,13 @@ impl Payload {
                 }
                 Ok(())
             },
+            (true, Payload::QuorumStoreInlineHybrid(_inline_batches, proof_with_status)) => {
+                for proof in proof_with_status.proof_with_data.proofs.iter() {
+                    proof.verify(validator)?;
+                }
+                // TODO: Do we need to make any checks to verify inline_batches?
```
Comment (bchocho) [inline consensus/consensus-types/src/common.rs:355]: Yes, we need to verify the inline batches to make sure the digest matches the txns. I think it's easiest if you actually use a `aptos_consensus::quorum_store::types::Batch` struct and call its verify function

Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:355]: I couldn't import aptos_consensus crate into consensus_types directory as it is causing cyclical dependency. I'll find a way to fix it.

File: consensus/src/quorum_store/proof_manager.rs [hunk 2]
```diff
@@ -88,23 +128,137 @@ impl ProofManager {
...
+                            excluded_batches.len(),
+                            proof_block.len(),
+                            inline_block.len()
+                        );
+                        // TODO: Need to calcuale max_txns_to_execute correctly here.
+                        Payload::QuorumStoreInlineHybrid(
+                            inline_block,
+                            ProofWithDataWithTxnLimit::new(
+                                ProofWithData::new(proof_block),
+                                Some(max_txns as usize),
```
Comment (bchocho) [inline consensus/src/quorum_store/proof_manager.rs:259]: This should be `None` and replaced later?

File: consensus/src/quorum_store/proof_manager.rs [hunk 3]
```diff
@@ -88,23 +128,137 @@ impl ProofManager {
...
-                let proof_block = self.proofs_for_consensus.pull_proofs(
-                    &excluded_batches,
-                    max_txns,
-                    max_bytes,
-                    return_non_full,
+                let (proof_block, proof_queue_not_fully_utilized) = self
+                    .proofs_for_consensus
+                    .pull_proofs(&excluded_batches, max_txns, max_bytes, return_non_full);
+
+                for excluded_batch in &excluded_batches {
```
Comment (bchocho) [inline consensus/src/quorum_store/proof_manager.rs:135]: I don't see the need to do all of this state cleanup here (from this line, down to removing expired batches). To make things easier to understand, can we be consistent with the proofs and remove on commit and expiration? (The only difference AFAICT should be removing on getting the full proof)

File: config/src/config/quorum_store_config.rs [hunk 4]
```diff
@@ -76,6 +77,7 @@ pub struct QuorumStoreConfig {
 impl Default for QuorumStoreConfig {
     fn default() -> QuorumStoreConfig {
         QuorumStoreConfig {
+            allow_batches_without_pos_in_proposal: true,
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply config/src/config/quorum_store_config.rs:80]: Sure. I will set this to false when landing.

File: consensus/consensus-types/src/common.rs [hunk 5]
```diff
@@ -137,6 +137,24 @@ impl ProofWithData {
...
     }
+
+    pub fn len(&self) -> usize {
+        self.proofs
+            .iter()
+            .map(|proof| proof.num_txns() as usize)
+            .sum()
+    }
+
+    pub fn size(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:148]: Done.

File: consensus/consensus-types/src/common.rs [hunk 6]
```diff
@@ -358,10 +426,29 @@ impl From<&Vec<&Payload>> for PayloadFilter {
...
+                        for proof in &proof_with_status.proof_with_data.proofs {
+                            exclude_proofs.insert(proof.info().clone());
+                        }
+                    },
+                    Payload::QuorumStoreInlineHybrid(inline_batches, proof_with_status) => {
+                        for proof in &proof_with_status.proof_with_data.proofs {
+                            exclude_proofs.insert(proof.info().clone());
+                        }
+                        for (batch_info, _) in inline_batches {
+                            // TODO: Couldn't verify whether the digest is correct in batch_info
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:445]: Then we need to verify somewhere else? So far, only the proof manager verifies the hash is correct before inserting the inline batch into the proposal. The other validators need to verify somewhere before signing the proposal.

File: consensus/consensus-types/src/common.rs (hunk 1 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs]: Done.

File: consensus/consensus-types/src/common.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bchocho) [reply consensus/consensus-types/src/common.rs]: Yup, we'll have to verify in Payload.verify. You have a TODO there:
> // TODO: Do we need to make any checks to verify inline_batches?

File: consensus/src/quorum_store/proof_manager.rs [hunk 7]
```diff
@@ -1,59 +1,94 @@
...
 pub enum ProofManagerCommand {
     ReceiveProofs(ProofOfStoreMsg),
+    ReceiveBatches(Vec<PersistedValue>),
     CommitNotification(u64, Vec<BatchInfo>),
     Shutdown(tokio::sync::oneshot::Sender<()>),
 }
 
 pub struct ProofManager {
     proofs_for_consensus: ProofQueue,
+    batches_without_proof_of_store: HashMap<BatchKey, Option<Vec<SignedTransaction>>>,
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:36]: Done.

File: consensus/src/quorum_store/proof_manager.rs (hunk 3 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs]: Earlier, the latency for 3000 TPS and 5000 TPS sweep tests was very high because we didn't cleanup these excluded batches. Cleaning up the excluded batches fixed the issue. These excluded batches won't be many. So, cleanup shouldn't take a lot of time hopefully.

File: consensus/src/quorum_store/proof_manager.rs (hunk 2 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:291]: Done.

File: config/src/config/quorum_store_config.rs [hunk 8]
```diff
@@ -107,6 +108,7 @@ impl Default for QuorumStoreConfig {
             // number of batch coordinators to handle QS batch messages, should be >= 1
             num_workers_for_remote_batches: 10,
             batch_buckets: DEFAULT_BUCKETS.to_vec(),
+            allow_batches_without_pos_in_proposal: true,
```
Comment (sitalkedia) [inline config/src/config/quorum_store_config.rs:111]: The default should be set to false right to maintain backward compatibility? Actually can you ensure all unit tests pass with this being default to true (so that the UTs exercise the newly added code path) but ensure to set this to false before landing?

File: consensus/src/quorum_store/proof_manager.rs (hunk 3 above)
Comment (bchocho) [reply consensus/src/quorum_store/proof_manager.rs]: We have to filter them, but we don't have to remove them

File: consensus/consensus-types/src/common.rs [hunk 9]
```diff
@@ -298,6 +353,21 @@ impl Payload {
                 }
                 Ok(())
             },
+            (true, Payload::QuorumStoreInlineHybrid(inline_batches, proof_with_status)) => {
+                for proof in proof_with_status.proof_with_data.proofs.iter() {
+                    proof.verify(validator)?;
+                }
+                for (batch, payload) in inline_batches.iter() {
+                    // TODO: Can cloning be avoided here?
+                    if BatchPayload::new(batch.author(), payload.clone()).hash() != *batch.digest()
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:362]: Possible to avoid cloning here?

File: consensus/consensus-types/src/proof_of_store.rs [hunk 10]
```diff
@@ -132,6 +132,10 @@ impl BatchInfo {
     pub fn gas_bucket_start(&self) -> u64 {
         self.gas_bucket_start
     }
+
+    pub fn is_expired(&self) -> bool {
+        self.expiration() < aptos_infallible::duration_since_epoch().as_micros() as u64
```
Comment (sitalkedia) [inline consensus/consensus-types/src/proof_of_store.rs:137]: This doesn't seem right, should be ` self.expiration() > aptos_infallible::duration_since_epoch().as_micros() as u64`?

File: consensus/src/quorum_store/proof_manager.rs [hunk 11]
```diff
@@ -1,59 +1,195 @@
...
+    pub fn remove_batch(&mut self, batch: &BatchInfo) {
+        self.batch_to_txns.remove(&BatchKey::from_info(batch));
+        if let Some(batch_tree) = self.author_to_batches.get_mut(&batch.author()) {
+            batch_tree.remove(&BatchSortKey::from_info(batch));
+        }
+    }
+
+    pub fn remove_batches(&mut self, batches: Vec<BatchInfo>) {
+        for batch in batches.iter() {
+            self.batch_to_txns.remove(&BatchKey::from_info(batch));
```
Comment (sitalkedia) [inline consensus/src/quorum_store/proof_manager.rs:74]: nit just call `remove_batch`

File: consensus/src/quorum_store/proof_manager.rs [hunk 12]
```diff
@@ -88,23 +226,67 @@ impl ProofManager {
                     PayloadFilter::InQuorumStore(proofs) => proofs,
                 };
 
-                let proof_block = self.proofs_for_consensus.pull_proofs(
-                    &excluded_batches,
-                    max_txns,
-                    max_bytes,
-                    return_non_full,
+                let (proof_block, proof_queue_not_fully_utilized) = self
```
Comment (sitalkedia) [inline consensus/src/quorum_store/proof_manager.rs:232]: nit: can you change the semantic to return if the proof queue is fully utilized rather than the negation? That way you can get rid of negations in the code block below and improve the readability.

File: consensus/src/quorum_store/proof_manager.rs [hunk 13]
```diff
@@ -161,6 +343,9 @@ impl ProofManager {
                             ProofManagerCommand::ReceiveProofs(proofs) => {
                                 self.receive_proofs(proofs.take());
                             },
+                            ProofManagerCommand::ReceiveBatches(batches) => {
```
Comment (sitalkedia) [inline consensus/src/quorum_store/proof_manager.rs:346]: @vusirikala - How do we think about not sending the batches eagerly from the Batch coordinator to the proof manager and instead reading from the DB whenever we need it? This has two advantages

1. We avoid the unnecessary cloning of the batches
2.  There is an attack surface currently that a rouge validator would blow up the memory. We can avoid that by storing it in DB instead.

@bchocho Thoughts?

File: consensus/src/quorum_store/proof_manager.rs (hunk 3 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs]: Done.

File: consensus/consensus-types/src/proof_of_store.rs (hunk 10 above)
Comment (vusirikala) [reply consensus/consensus-types/src/proof_of_store.rs:137]: A batch is expired if the current time is higher than self.expiration() right?

File: consensus/src/quorum_store/proof_manager.rs (hunk 11 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:74]: Done.

File: consensus/consensus-types/src/common.rs [hunk 14]
```diff
@@ -235,6 +272,12 @@ impl Payload {
                     || (proof_with_status.max_txns_to_execute.is_some()
                         && proof_with_status.max_txns_to_execute.unwrap() == 0)
             },
+            Payload::QuorumStoreInlineHybrid(inline_batches, proof_with_status) => {
+                (proof_with_status.max_txns_to_execute.is_some()
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:281]: Done.

File: consensus/consensus-types/src/proof_of_store.rs (hunk 10 above)
Comment (sitalkedia) [reply consensus/consensus-types/src/proof_of_store.rs:137]: Duh, I am dumb  😅

File: consensus/consensus-types/src/common.rs [hunk 15]
```diff
@@ -177,6 +195,12 @@ pub enum Payload {
     DirectMempool(Vec<SignedTransaction>),
     InQuorumStore(ProofWithData),
     InQuorumStoreWithLimit(ProofWithDataWithTxnLimit),
+    // TODO: This is weird. `max_txns_to_execute` for the overall payload is embedded insie `ProofWithData`.
+    // Is that fine or should we change it?
+    QuorumStoreInlineHybrid(
+        Vec<(BatchInfo, Vec<SignedTransaction>)>,
+        ProofWithDataWithTxnLimit,
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:207]: Done.

File: config/src/config/quorum_store_config.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vusirikala) [reply config/src/config/quorum_store_config.rs]: Yes. I will flip the flag before landing.

File: config/src/config/quorum_store_config.rs (hunk 8 above)
Comment (vusirikala) [reply config/src/config/quorum_store_config.rs:111]: Yes. I will flip the flag before landing.

File: consensus/src/quorum_store/proof_manager.rs (hunk 12 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:225]: Done.

File: consensus/src/liveness/proposal_generator.rs [hunk 16]
```diff
@@ -335,6 +335,9 @@ impl ProposalGenerator {
                     self.quorum_store_poll_time.saturating_sub(proposal_delay),
                     max_block_txns,
                     max_block_bytes,
+                    // TODO: Set max_inline_txns and max_inline_bytes correctly
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/liveness/proposal_generator.rs:338]: Done.

File: consensus/src/payload_client/mixed.rs [hunk 17]
```diff
@@ -150,6 +154,8 @@ async fn mixed_payload_client_should_prioritize_validator_txns() {
             Duration::from_secs(1), // max_poll_time
             99,                     // max_items
             1048576,                // size limit: 1MB
+            50,
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/payload_client/mixed.rs:157]: Done.

File: consensus/consensus-types/src/common.rs (hunk 9 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:367]: I'm getting an error if I don't clone though. Not sure if there is a better way.

Comment (sitalkedia) [reply consensus/consensus-types/src/common.rs:367]: I see. We can leave it this way in that case.

File: consensus/src/quorum_store/batch_store.rs [hunk 18]
```diff
@@ -197,7 +197,7 @@ impl BatchStore {
     // Note: holds db_cache entry lock (due to DashMap), while accessing peer_quota
     // DashMap. Hence, peer_quota reference should never be held while accessing the
     // db_cache to avoid the deadlock (if needed, order is db_cache, then peer_quota).
-    pub(crate) fn insert_to_cache(&self, mut value: PersistedValue) -> anyhow::Result<bool> {
+    pub(crate) fn insert_to_cache(&self, value: &PersistedValue) -> anyhow::Result<bool> {
```
Comment (sitalkedia) [inline consensus/src/quorum_store/batch_store.rs:200]: Can you clarify why did you need to change this to a reference and hence require an extra clone?

File: consensus/src/quorum_store/batch_store.rs [hunk 19]
```diff
@@ -300,7 +301,7 @@ impl BatchStore {
     }
 
     fn persist_inner(&self, persist_request: PersistedValue) -> Option<SignedBatchInfo> {
-        match self.save(persist_request.clone()) {
```
Comment (vusirikala) [inline consensus/src/quorum_store/batch_store.rs:303]: @sitalkedia I made the above change in `insert_to_cache` function to eliminate this clone() operation. Earlier, we were cloning the persist request all the time and then potentially replace the payload with None. Now, I'm cloning the persist request only when it is required.

File: consensus/src/quorum_store/batch_coordinator.rs [hunk 20]
```diff
@@ -59,8 +67,19 @@ impl BatchCoordinator {
...
+            // Sending only small batches to proof manager as large batches can't be
+            // included in the inline transactions anyway.
+            let batches = persist_requests
+                .iter()
+                .map(|persisted_value| persisted_value.batch_info().clone())
+                .collect();
+            let _ = sender_to_proof_manager
+                .send(ProofManagerCommand::ReceiveBatches(batches))
+                .await;
             let signed_batch_infos = batch_store.persist(persist_requests);
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/batch_coordinator.rs:83]: Done.

**Comment by sitalkedia** [comment]
@vusirikala - Can you rebase and run all unit tests and smoke tests with the feature flag on?

File: consensus/src/quorum_store/proof_manager.rs [hunk 21]
```diff
@@ -1,59 +1,175 @@
...
+                if full {
+                    return false;
+                }
+                if let Some((_sort_key, batch)) = iter.next() {
+                    if excluded_batches.contains(batch) {
+                        true
+                    } else if num_txns + batch.num_txns() <= max_txns
+                        && num_bytes + batch.num_bytes() <= max_bytes
+                    {
+                        if let Ok(mut persisted_value) =
```
Comment (bchocho) [inline consensus/src/quorum_store/proof_manager.rs:106]: Can we log warn! if it's not found? Shouldn't really happen often, so if does maybe having logs will help.

File: consensus/src/quorum_store/batch_coordinator.rs [hunk 22]
```diff
@@ -59,8 +67,19 @@ impl BatchCoordinator {
 
         let batch_store = self.batch_store.clone();
         let network_sender = self.network_sender.clone();
+        let sender_to_proof_manager = self.sender_to_proof_manager.clone();
         tokio::spawn(async move {
             let peer_id = persist_requests[0].author();
+            // TODO: Is it fine to ignore the result of send here?
+            // Sending only small batches to proof manager as large batches can't be
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/batch_coordinator.rs:74]: Done.

File: consensus/src/quorum_store/proof_manager.rs (hunk 21 above)
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:106]: Done.

File: consensus/consensus-types/src/common.rs [hunk 23]
```diff
@@ -233,18 +279,90 @@ impl Payload {
...
+            (
+                Payload::QuorumStoreInlineHybrid(b1, p1, m1),
+                Payload::QuorumStoreInlineHybrid(b2, p2, m2),
+            ) => {
+                let mut b3 = b1;
+                b3.extend(b2);
+                let mut p3 = p1;
+                p3.extend(p2);
+                // TODO: What's the right logic here?
+                let m3 = if m1.is_none() {
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:321]: nit: Can you change this to `match` statement? It would be more readable and cleaner that way rather than a bunch of if-else. Something like this should do 

```
match (m1, m2) {
 None,  m2 => m2,
 Some(m1), Some(m2) =>
.......

```

File: consensus/consensus-types/src/common.rs [hunk 24]
```diff
@@ -233,18 +279,90 @@ impl Payload {
...
+            },
+            (
+                Payload::QuorumStoreInlineHybrid(b1, p1, m1),
+                Payload::QuorumStoreInlineHybrid(b2, p2, m2),
+            ) => {
+                let mut b3 = b1;
+                b3.extend(b2);
+                let mut p3 = p1;
+                p3.extend(p2);
+                // TODO: What's the right logic here?
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:320]: Is this TODO still applicable?

File: consensus/consensus-types/src/common.rs [hunk 25]
```diff
@@ -233,18 +279,90 @@ impl Payload {
...
+            },
+            (Payload::QuorumStoreInlineHybrid(b1, p1, m1), Payload::InQuorumStore(p2)) => {
+                // TODO: How to update m1?
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::QuorumStoreInlineHybrid(b1, p3, m1)
+            },
+            (Payload::QuorumStoreInlineHybrid(b1, p1, m1), Payload::InQuorumStoreWithLimit(p2)) => {
+                // TODO: What's the right logic here?
+                let m3 = if m1.is_none() {
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:338]: Replace if-else with match statement.

File: consensus/consensus-types/src/common.rs [hunk 26]
```diff
@@ -233,18 +279,90 @@ impl Payload {
...
+                Payload::QuorumStoreInlineHybrid(b1, p3, m3)
+            },
+            (Payload::InQuorumStore(p1), Payload::QuorumStoreInlineHybrid(b2, p2, m2)) => {
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::QuorumStoreInlineHybrid(b2, p3, m2)
+            },
+            (Payload::InQuorumStoreWithLimit(p1), Payload::QuorumStoreInlineHybrid(b2, p2, m2)) => {
+                // TODO: What's the right logic here?
+                let m3 = if p1.max_txns_to_execute.is_none() {
```
Comment (sitalkedia) [inline consensus/consensus-types/src/common.rs:356]: Same as above. Also see if its possible to refactor the logic into a common function.

File: consensus/consensus-types/src/common.rs (hunk 23 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:321]: Done.

File: consensus/consensus-types/src/common.rs (hunk 25 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:338]: Done.

File: consensus/consensus-types/src/common.rs (hunk 26 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs]: Refactored using a common function.

File: consensus/consensus-types/src/common.rs (hunk 24 above)
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs:328]: Yes. I'm not 100% confident about the right logic here.

**Review by sitalkedia — APPROVED** [review]
Thank you for working on this - great work!

File: consensus/consensus-types/src/common.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/consensus-types/src/common.rs]: The logic is actually is already in a function.

File: consensus/src/quorum_store/proof_manager.rs [hunk 27]
```diff
@@ -63,7 +179,9 @@ impl ProofManager {
             "QS: got clean request from execution at block timestamp {}",
             block_timestamp
         );
-
+        for batch in &batches {
+            self.batch_queue.remove_batch(batch);
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/proof_manager.rs:186]: Done.

**Review by bchocho — APPROVED** [review]
